man/canto.1
.TH Canto 1 "MAN_DATE" "Version MAN_VERSION" "Canto"

.SH NAME
Canto \- An ncurses RSS reader
.SH DESCRIPTION
Canto is an RSS reader built to be flexible and highly customizable on top of python.

.SH GETTING STARTED
Canto is the client, separate from the fetcher Canto-fetch. In order to update you must invoke canto-fetch on a regular basis. Usually the way to do that is to add a line into your crontab like this:

* * * * * canto-fetch

If you're not a fan of cron, then you can add `canto-fetch \-b` to your startup
scripts to have canto-fetch run as a daemon on its own.

After you have your configuration file generated, just run `canto` and it will update for you, if canto-fetch hasn't been run already.

.SH COMMAND LINE USAGE
.TP
\-h/--help
Simple help

.TP
\-v/--version
Print version

.TP
\-u/--update
Update feeds before launching client.

.TP
\-l/--list
List all configured feeds

.TP
\-a/--checkall
Print number of new items.

.TP
\-n/--checknew [feed]
Print number of new items in feed.

.TP
\-o/--opml
Print feeds to stdout as OPML file.

.TP
\-i/--import [path]
Import feeds from OPML file to your configuration.

.TP
\-r/--url [URL]
Add feed at URL to your configuration.

.TP
\-t/--tag [tag]
Set tag for feed added with \-r

.TP
\-D/--dir [path]
Set the configuration directory (default: ~/.canto/)

.TP
\-C/--conf [path]
Set the configuration file (default: ~/.canto/conf)

.TP
\-L/--log [path]
Set the log file (default: ~/.canto/log)

.TP
\-F/--fdir [path]
Set the feed directory (default: ~/.canto/feeds)

.TP
\-S /--sdir [PATH]
Set the path to execurl scripts (default ~/.canto/scripts/)

.SH INTERNAL USAGE
Within the program you can use the following (default) keys.
These can be changed in your configuration file by using the
"key" configuration option.

.TP
UP / DOWN or k / j
Select previous or next item. (next_item) (prev_item)

.TP
PGUP / PGDOWN or o / l
Goto previous or next tag. (next_tag) (prev_tag)

.TP
RIGHT / LEFT
Set item unread or read (just_unread) (just_read)

.TP
[ / ]
Cycle through defined filters (prev_filter) (next_filter)

.TP
{ / }
Cycle through defined tag filters (prev_feed_filter) (next_feed_filter)

.TP
\- / =
Cycle through defined tag sorts (prev_tag_sort) (next_tag_sort)

.TP
< / >
Cycle through defined tag sets (prev_tagset) (next_tagset)

.TP
:
Goto a specific tag (order of the config) (goto_tag)

.TP
;
Goto a specific visible tag (goto_reltag)

.TP
TAB
Switch focus between list and reader (only useful with dedicated reader space)
    (switch)

.TP
h
Display this man page. (help)

.TP
Space
Read a story (reader)

.TP
g
Use the defined browser to goto the item's URL (goto)

.TP
C / V
Collapse/Uncollapse all tags (set_collapse_all) (unset_collapse_all)

.TP
c
Collapse/Uncollapse current tag (toggle_collapse_tag)

.TP
f
Inline search (inline_search)

.TP
m
Toggle item marked/unmark (toggle_mark)

.TP
M
Unmark all items (all_unmarked)

.TP
n / p
Goto next/previous marked item (next_mark) (prev_mark)

.TP
, / .
Goto next/previous unread item (next_unread) (prev_unread)

.TP
r / u
Mark tag read/unread (tag_read) (tag_unread)

.TP
R / U
Mark all read/unread (all_read) (all_unread)

.TP
Ctrl+R
Refresh feeds (force_update)

.TP
Ctrl+L
Redraw screen (refresh)

.TP
\\
Restart Canto (restart)

.TP
q
Quit Canto (quit)

.SH READER USAGE
Inside the reader, there are a number of different keys. These can be changed with the "reader_key" configuration option.

.TP
UP / DOWN or k / j
Scroll up/down if content off screen (scroll_up) (scroll_down)

.TP
n / p
goto next/previous item without closing reader

.TP
l
Enumerate links (toggle_show_links)

.TP
g
Choose a link to goto (goto)

.SH CONFIGURATION
The ~/.canto/conf.py file is where all of the configuration is. You can start by
reading http://codezen.org/canto/config . If you're updating from <= 0.6.x then you
should read http://codezen.org/canto/config/#upgrading-from-06x .

.SH FILES
.TP
.I ~/.canto/conf.py
Main configuration file. ~/.canto/conf (without the extension) is also checked
for compatibility.

.TP
.I ~/.canto/log
Everyday log file.

.TP
.I ~/.canto/fetchlog
Canto-fetch log file.

.TP
.I ~/.canto/feeds/
This is the directory where the stories are recorded.

.SH BUGS
I'm sure there are some. If you run into a bug (a crash or bad behavior), then
send please report it. Any of the methods described in
http://codezen.org/canto/contact are acceptable. Also, please include your
configuration and log files with the report.

.SH HOMEPAGE
http://codezen.org/canto

.SH AUTHOR
Jack Miller <jack@codezen.org>


ChangeLog
Canto 0.7.9 ChangeLog
* Fix fresh install sans system feedparser
* Make canto-inspect use builtin feedparser

Canto 0.7.8 ChangeLog
* Fix feed exception encoding problem.
* Fix occasional zombies / extra pids floating around.
* Import feedparser into source tree.
* Add -s/--sysfp flag to canto-fetch to fall back on system feedparser
* Render improvements by honoring declared content types

Canto 0.7.7 ChangeLog
* Fix harmless widecurse.c warning
* Fix going to locale incompatible URLs
* Fix Python 2.6.5 weirdness
* Fix enclosure parsing/display exception
* Fix set_{filter,tag_filter,sort}(None)
* Improved config validation for gui options
* Work harder to maintain selections through updates
* Make cursor behavior more flexible
* Change default cursor behavior
* Documentation tweaks (thanks acoolon)

Canto 0.7.6 ChangeLog
* Fix non-consecutive sorts
* Fix multi-tag sorts
* Fix change_feed
* Fix possible get_text exception
* Replace multiprocessing functionality
* Worker process initiated on demand
* Add "restart" keybind (|)
* Add wget_link reader_key
* Add highlight_word drawing hook
* Add "setup.py uninstall" command
* Minor c-f cleanup

Canto 0.7.5 ChangeLog
* Fix some reader inconsistencies with reader keys moving selections
* Wrap some harmless, rare curses exceptions
* Workaround messed up SIGCHLD handling in multiprocessing
* Doc updates

Canto 0.7.4 ChangeLog
* Correct some overlooked 0.7.3 stuff =P

Canto 0.7.3 ChangeLog
* Fix various update issues on long-running clients
* Fix sluggish reader link toggling
* Fix worker signals (and ^C as a side-effect)
* Fix reader n/p keys not setting items read
* Fix double quotes in programmatically added main tags
* Fix shadows on horrendously broken feeds
* Fix all-filtered stub
* Refix hard-filters (??)
* Minor cleanups
* Documentation clarifications

Canto 0.7.2 ChangeLog
* Fix some precache troubles with aggregate filters / reverse
* Restore feed-order without sort
* Startup cleanups

Canto 0.7.1 ChangeLog
* Fix hard (feed) filters
* Fix keyword escaping for non-regex searches
* Fix items with totally undefined titles
* Fix fetchlog header from arg refactor
* Ignore some exceptions cause by multiprocessing
* Minor doc tweaks

Canto 0.7.0 ChangeLog
* Convert to multiprocessing worker slave (huge performance)
* Vast memory improvement (esp. for large lists)
* Large scale refactor of *lots* of code
* Better code documentation
* Better site documentation
* Partially validating configuration code
* Partial test framework (to be added to as bugs arise)
* Added configurable update triggers
* Added `never_discard()` to keep certain items indefinitely.
* Added SIGUSR2 signal to output debug backtrace.
* Added state_change_hook
* Added `canto-inspect`, a simple wrapper for examining feed internals
* Added no-content stub for unfetchable feeds to avoid trying to update broken
	URLS repeatedly.
* Added `add_info` extra function for adding content to the reader
* new_hook now enforced by canto-fetch
* Ignore keep settings lower than the number of items in a feed
* tags variably now implicitly set to sane default
* "reader" keybind now doesn't set item read (coupled with "just_read" for
	default)
* Filters and Sorts now all subclass Filter and Sort class
* Accept `conf.py` as well as `conf` for config name.
* Fix double enforcement of rates (client more responsive to fetch updates)
* Fix blank titles
* Fix runhere.sh killing c-f

Canto 0.6.13 ChangeLog
* Fix drawing regression

Canto 0.6.12 ChangeLog
* Fix tag crash
* Fix strange character weirdness
* Fix/Improve HTML parser
* Try UTF-8 for config before chardet

Canto 0.6.11 ChangeLog
* Fix OPML handling for OPML without text attribute
* Exception clean ups
* Doc clean ups
* canto.extra additions

Canto 0.6.10 ChangeLog
* Make HTML parser more resistant to broken HTML
* Fix minor exception in c-f thread
* Finally make exceptions play nice with ncurses

Canto 0.6.9 ChangeLog
* Fix setup.py generating null bytes in const.py
* Add 30 second timeout to canto-fetch
* Make selection data persist for hooks / filters
* Unset signals before exit (avoid shell garbage)
* Set User-Agent to Canto/x.y.z (fixes some 403'd feeds)
* Fix multiple c-f subtle corruption bug
* Sync docs now that site runs out of git.

Canto 0.6.8 ChangeLog
* Fix set_tag_sort(None)
* Fix miscount of items when all filtered
* Reader keybinds now passed reader object (like Gui keybinds)
* Cache overused locale query (crazy speedup on that)
* Convert setup.py to pure python (no more sed scripts)

Canto 0.6.7 ChangeLog
* Make filter syntax uniform (add filters now work with canto/extra filters)
* Fix add_tag() without sorts= set (typo)

Canto 0.6.6 ChangeLog
* Make setup.py sed scripts BSD compatible
* Even basic print statements now obey locale.preferredencoding()
* All input command-line input is unicode'd() (-n fix)

Canto 0.6.5 ChangeLog
* Fix multiple identical main tag weirdness
* Fix curses crash with TERM misset

canto/canto.py
# -*- coding: utf-8 -*-

#Canto - ncurses RSS reader
#   Copyright (C) 2008 Jack Miller <jack@codezen.org>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License version 2 as 
#   published by the Free Software Foundation.

# The Main object encompasses a running instance of Canto. It can be divided
# into a number of parts.
#    init
#        if called as canto-fetch jump to canto_fetch.main
#        parse config
#        parse canto specific arguments
#        start work thread
#        load the initial feed data (from thread)
#        do one-off operation flags (like -a, -n, -o, -i, etc.)
#        do basic curses init
#        instantiate Gui class
#    main loop
#        check that Gui is still alive
#        if we're waiting for a process, sleep
#        check for input
#            if no input, check on threads
#                if work done, update screen
#            if input, pass to Gui and interpret return
#                if return implies update, queue up work for thread

from thread import ThreadHandler
from utility import Cycle
from cfg.base import get_cfg
from const import *
from gui import Gui

import canto_fetch
import utility
import args
import tag

import traceback
import signal
import locale
import curses
import time
import sys
import os

class Main():
    def __init__(self):
        signal.signal(signal.SIGUSR2, self.debug_out)

        # Let locale figure itself out
        locale.setlocale(locale.LC_ALL, "")
        enc = locale.getpreferredencoding()

        # If we're canto-fetch, jump to that main function
        if sys.argv[0].endswith("canto-fetch"):
            canto_fetch.main(enc)

        # Parse arguments that canto shares with canto-fetch, return
        # a lot of file locations and an optlist that will contain the
        # parsed, but yet unused canto specific arguments.

        conf_dir, log_file, conf_file, feed_dir, script_dir, optlist =\
                args.parse_common_args(enc,
                    "hvulaor:t:i:n:",
                    ["help","version","update","list","checkall","opml",
                        "import=","url=","checknew=","tag="])

        # Instantiate the config and start the log.
        try :
            self.cfg = get_cfg(conf_file, log_file, feed_dir, script_dir)
            self.cfg.parse()
        except :
            traceback.print_exc()
            sys.exit(-1)

        self.cfg.log("Canto v %s (%s)" % \
                ("%d.%d.%d" % VERSION_TUPLE, GIT_SHA), "w")
        self.cfg.log("Time: %s" % time.asctime())
        self.cfg.log("Config parsed successfully.")

        # Default arguments.
        flags = 0 
        feed_ct = None
        opml_file = None
        url = None
        newtag = None

        # Note that every single flag that takes an argument has its
        # argument converted to unicode. Saves a lot of bullshit later.

        for opt, arg in optlist :
            if opt in ["-u","--update"] :
                flags |= UPDATE_FIRST
            elif opt in ["-n","--checknew"] :
                flags |= CHECK_NEW
                feed_ct = unicode(arg, enc, "ignore")
            elif opt in ["-a","--checkall"] :
                flags |= CHECK_NEW
            elif opt in ["-l","--list"] :
                flags |= FEED_LIST
            elif opt in ["-o","--opml"] :
                flags |= OUT_OPML
            elif opt in ["-i","--import"] :
                flags |= IN_OPML
                opml_file = unicode(arg, enc, "ignore")
            elif opt in ["-r","--url"] :
                flags |= IN_URL
                url = unicode(arg, enc, "ignore")
            elif opt in ["-t","--tag"] :
                newtag = unicode(arg, enc, "ignore")

        # Import flags harness the same functions as their config
        # based counterparts, source_opml and source_url.
        # XXX THIS SHIT IS BROKEN

        if flags & IN_OPML:
            self.cfg.source_opml(opml_file, append=True)
            print "OPML imported."

        if flags & IN_URL:
            self.cfg.source_url(url, append=True, tag=newtag)
            print "URL added."

        # All import options should terminate.

        if flags & (IN_OPML + IN_URL):
            sys.exit(0)

        # If self.cfg had to generate a config, make sure we
        # update first.

        if self.cfg.no_conf:
            self.cfg.log("Conf was auto-generated, adding -u")
            flags |= UPDATE_FIRST

        if flags & UPDATE_FIRST:
            self.cfg.log("Pausing to update...")
            canto_fetch.run(self.cfg, True, True)

        # Detect if there are any new feeds by whether their
        # set path exists. If not, run canto-fetch but don't
        # force it, so canto-fetch intelligently updates.

        for i,f in enumerate(self.cfg.feeds) :
            if not os.path.exists(f.path):
                self.cfg.log("Detected unfetched feed: %s." % f.URL)
                canto_fetch.run(self.cfg, True, False)

                #Still no go?
                if not os.path.exists(f.path):
                    self.cfg.log("Failed to fetch %s, removing" % f.URL)
                    self.cfg.feeds[i] = None
                else:
                    self.cfg.log("Fetched.\n")
                break

        # Collapse the feed array, if we had to remove some unfetchables.
        self.cfg.feeds = filter(lambda x: x != None, self.cfg.feeds)

        self.new = []
        self.old = []
        self.th = ThreadHandler()

        # Force an update from disk by queueing a work item for each thread.
        # At this point, we just want to do the portion of the update where the
        # disk is read, so THREAD_UPDATE is used.

        self.cfg.log("Populating feeds...")
        for f in self.cfg.feeds:
            self.th.update.put((self.cfg, f, [], THREAD_UPDATE))

        # Wait for all of the disk updates to be done.
        self.th.update.join()

        # This chunk of code takes any base tags that inadvertantly conflict
        # (i.e. weren't explicitly set by the user) and resolves them into
        # Tag, Tag (2), Tag (3), etc.

        # This may seem like paranoia, but half-assed feed generators that use
        # default feed titles shouldn't break Canto.

        base_tags = {}
        for f in [x for x in self.cfg.feeds if \
                x.base_set and not x.base_explicit]:
            otag = f.tags[0]
            if f.tags[0] in base_tags:
                base_tags[otag] += 1

                # We check each tag is unique, even if we're generating a new
                # one so that if a user defines "Tag, Tag, Tag (2)", it resolves
                # to "Tag, Tag (3), Tag (2)"

                while f.tags[0] + (" (%d)" % base_tags[otag]) in base_tags:
                    base_tags[otag] += 1
                f.tags[0] += " (%d)" % base_tags[otag]

                # Remove original tag from all stories in the feed
                for s in f:
                    s.unset(otag)
                    s.set(f.tags[0])
            else:
                base_tags[f.tags[0]] = 1

        # Now that the tags have all been straightened out, validate the config.
        # Making sure the tags are unique before validation is important because
        # part of validation is the actual creation of Tag() objects.

        try:
            self.cfg.validate()
        except Exception, err:
            print err
            sys.exit(0)

        # Print out a feed list
        if flags & FEED_LIST:
            for f in self.cfg.feeds:
                print f.tags[0].encode(enc, "ignore")
            sys.exit(0)

        # This could probably be done better, or more officially
        # with some XML library, but for now, print is working
        # fairly well.

        if flags & OUT_OPML:
            self.cfg.log("Outputting OPML")
            print """<opml version="1.0">"""
            print """<body>"""
            for feed in self.cfg.feeds:
                if "atom" in feed.ufp["version"]:
                    t = "pie"
                else:
                    t = "rss"

                print """\t<outline text="%s" xmlUrl="%s" type="%s" />""" %\
                        (feed.tags[0].encode(enc, "ignore"), feed.URL, t)

            print """</body>"""
            print """</opml>"""
            sys.exit(0)

        # After this point, we know that all further operation is going to
        # require tags to be populated, we queue up the latter half of the
        # work, to actually filter and sort the items.

        for f in self.cfg.feeds:
            self.th.update.put((self.cfg, f, [], THREAD_FILTER))

        # Handle -a/-n flags (print number of new items)
        # XXX THIS SHIT IS BROKEN
        if flags & CHECK_NEW:
            if not feed_ct:
                feed_ct = "*"

            # We get counts by using the Tag() class directly.
            check_tag = tag.Tag(self.cfg, Cycle([[None]]),\
                    Cycle([None]), feed_ct)
            check_tag.extend(self.stories)
            print check_tag.unread
            sys.exit(0)

        # At this point we know that we're going to actually launch
        # the client, so we fire up ncurses and add the screen
        # information to our Cfg().

        self.cfg.stdscr = curses.initscr()
        self.cfg.stdscr.nodelay(1)

        # curs_set can return ERR, we shouldn't care
        try:
                curses.curs_set(0)
        except:
                pass

        # if any of these mess up though, the rest of the
        # the operation is suspect, so die.
        try:
                curses.noecho()
                curses.start_color()
                curses.use_default_colors()
        except:
                self.cfg.log("Unable to init curses, bailing")
                self.done()

        self.sigusr = 0
        self.resize = 0
        self.alarmed = 1
        self.ticks = 1

        self.cfg.height, self.cfg.width = self.cfg.stdscr.getmaxyx()

        # Init colors
        for i, (fg, bg) in enumerate(self.cfg.colors):
            curses.init_pair(i + 1, fg, bg)

        self.cfg.log("Curses initialized.")
    
        # Instantiate the base Gui class
        self.gui = Gui(self.cfg, self.cfg.tags.cur())

        self.cfg.log("GUI initialized.")

        # Signal handling
        signal.signal(signal.SIGWINCH, self.winch)
        signal.signal(signal.SIGALRM, self.alarm)
        signal.signal(signal.SIGCHLD, self.chld)
        signal.signal(signal.SIGINT, self.done)
        signal.signal(signal.SIGUSR1, self.sigusr)

        self.cfg.log("Signals set.")
        self.estring = None

        # The main loop is wrapped in one huge try...except so that no matter
        # what exception is thrown, we can clean up curses and exit without
        # shitting all over the terminal.

        try:
            # Initial draw of the screen
            self.refresh()

            # Main program loop, terminated when all handlers have
            # deregistered / exited.

            self.cfg.log("Beginning main loop.")

            while 1:
                # Clear the key
                t = None

                # Gui is none if a key returned EXIT
                if not self.gui:
                    break

                # If we've spawned a text-based process (i.e. elinks), then just
                # pause and wait to be awakened by SIGCHLD

                if self.cfg.wait_for_pid:
                    signal.pause()

                # Tick when SIGALRM is received.
                if self.alarmed:
                    self.alarmed = 0
                    self.tick()

                # Deferred update from signal
                if self.sigusr:
                    self.sigusr = 0
                    if "signal" in self.cfg.triggers:
                        self.update()

                # Get the key
                k = self.cfg.stdscr.getch()

                # KEY_RESIZE is the only key not propagated, to
                # keep users from rebinding it and crashing.

                if k == curses.KEY_RESIZE or self.resize:
                    self.resize = 0
                    self.refresh()
                    continue

                # No input, time to check on the threads.

                if k == -1:

                    # Make sure we don't pin the CPU, so if there's no input and
                    # no waiting updates, sleep for awhile.

                    if self.th.updated.empty():
                        time.sleep(0.01)
                    else:
                        r = self.th.updated.get()
                        if r:
                            self.gui.alarm(r[0], r[1])
                            self.gui.draw_elements()
                    continue

                # Handle Meta pairs
                elif k == 195:
                    k2 = self.cfg.stdscr.getch()
                    if k2 >= 64:
                        t = (k2 - 64, 1)
                    else:
                        t = (k, 0)

                # Just a normal key-press
                else:
                    t = (k, 0)

                # Key resolves a keypress tuple into a list of actions
                actions = self.gui.key(t)

                # Actions are executed in order, and each return code 
                # is handled in order.

                for a in actions:
                    r = self.gui.action(a)
                    if r == REFRESH_ALL:
                        self.refresh()
                    elif r == UPDATE:
                        self.update()
                    elif r in [REFILTER, RETAG]:
                        self.th.flush()
                        self.update(1)
                    elif r == TFILTER:
                        # Tag filters shouldn't perform a full update, so we map
                        # the relevant tag to all of the feeds that include that
                        # tag and update them.
                        t = self.gui.sel["tag"]
                        ufds = [ f for f in self.cfg.feeds\
                                if t.tag in f.tags]
                        t.clear()

                        # Perform the update on necessary feeds, we give them
                        # top priority so that even if the interface isn't fully
                        # updated, a tag filter change results in a noticeable
                        # change ASAP.
                        self.update(1, ufds, 1)
                    elif r == REDRAW_ALL:
                        self.gui.draw_elements()
                    elif r == EXIT:
                        self.gui = None
                        break

        except Exception:
            self.estring = traceback.format_exc()
        except KeyboardInterrupt:
            pass

        self.done()

    def done(self, a=None, b=None):
        # Unset signals.
        for s in [signal.SIGALRM, signal.SIGWINCH,
                signal.SIGCHLD, signal.SIGINT]:
            signal.signal(s, signal.SIG_IGN)

        # Kill the message log
        self.cfg.msg = None

        # Kill curses
        curses.endwin()

        self.cfg.log("Curses done.")

        # If there was an exception, nicely print it out.
        if self.estring:
            self.cfg.log("\nEXCEPTION:")
            self.cfg.log(self.estring)
            print "Canto exited on an exception.\n"
            print self.estring
            print "Please report this bug. Send your logfile " +\
                "(%s) to jack@codezen.org" % self.cfg.log_file

        # Flush the work thread to make sure no updates are going on.
        self.th.flush(0)

        # Make sure we leave the on-disk presence constant
        for feed in self.cfg.feeds:
            while feed.changed():
                feed.todisk()

        self.cfg.log("Flushed to disk.")
        sys.exit(0)

    # For the most part, it's smart to avoid doing anything but set a flag in an
    # signal handler. CHLD is an exception because the only case in which we do
    # anymore work than just acknowledging the process is dead is when
    # wait_for_pid is set and, in this case the main loop is paused anyway.

    def chld(self, a=None, b=None):
        # I'm not sure why, but SIGCHLD gets called and occasionally, os.wait()
        # then complains about there being no waiting processes.
        try:
            pid,none = os.wait()
        except:
            return

        # If the interface is waiting on this pid to be done,
        # reset the signal and simulate a resize to make sure the window
        # information is still fresh.

        if self.cfg.wait_for_pid == pid:
            self.cfg.wait_for_pid = 0
            signal.signal(signal.SIGALRM, self.alarm)
            signal.signal(signal.SIGWINCH, self.winch)
            self.resize = 1

    # Back to better practices. =)

    def winch(self, a=None, b=None):
        self.resize = 1

    def alarm(self, a=None, b=None):
        self.alarmed = 1

    def sigusr(self, a, b):
        self.sigusr = 1

    # Tick decrements two timers. One for a possible update (if "interval" is a
    # valid update trigger), and one for the message box at the bottom of the
    # interface so that messages don't persist for very long.

    def tick(self, refilter=0):
        # Possible update tick
        self.ticks -= 1
        if self.ticks <= 0:
            if "interval" in self.cfg.triggers:
                for f in self.cfg.feeds:
                    f.time -= 1
                self.update(0, [f for f in self.cfg.feeds if f.time < 0])
            self.ticks = 60

        # Message tick
        self.cfg.msg_tick -= 1
        if self.cfg.msg_tick == 0:
            self.cfg.message(self.cfg.status(self.cfg), 1)

        signal.alarm(1)

    # Update is where the work is queued up for the work thread.
    def update(self, refilter = 0, iter = None, priority = 0):
        old = []

        # Default to updating all feeds
        if not iter:
            iter = self.cfg.feeds

        for f in iter:

            # If we're not refiltering, compare against the current state of the
            # feed, otherwise we count on the tags being empty.

            if not refilter:
                old = f[:]

            # Priority items, like tag refilter updates are moved to the front
            # of the queue, others are queued as normal.

            if priority:
                self.th.update.put_next((self.cfg, f, old, THREAD_BOTH))
            else:
                self.th.update.put((self.cfg, f, old, THREAD_BOTH))

    # Refresh should only be called when it's possible that the screen has
    # changed shape. 

    def refresh(self):
        # Get new self.cfg.{height, width}
        curses.endwin()
        self.cfg.stdscr.touchwin()
        self.cfg.stdscr.refresh()
        self.cfg.stdscr.keypad(1)

        self.cfg.height, self.cfg.width = self.cfg.stdscr.getmaxyx()

        # If there's a resize hook, execute it.
        if self.cfg.resize_hook:
            self.cfg.resize_hook(self.cfg)

        # Make sure we've got a minimum for columns and reader_lines
        self.cfg.columns = max(self.cfg.columns, 1)
        self.cfg.reader_lines = max(self.cfg.reader_lines, 3)

        # Adjust gui_height to compensate for the message at the bottom.
        self.cfg.gui_height = self.cfg.height - self.cfg.msg_height
        self.cfg.gui_width = self.cfg.width

        # Now we interpret the reader_orientation setting from the config to
        # shape the reader area, and adjust other height / width settings
        # accordingly.

        # XXX This logic could be cleaned up...

        if self.cfg.reader_orientation == "top":
            self.cfg.gui_height -= self.cfg.reader_lines
            self.cfg.gui_top = self.cfg.reader_lines
        elif self.cfg.reader_orientation == "bottom":
            self.cfg.gui_height -= self.cfg.reader_lines
        elif self.cfg.reader_orientation == "left":
            self.cfg.gui_width -= self.cfg.reader_lines
            self.cfg.gui_right = self.cfg.reader_lines
        elif self.cfg.reader_orientation == "right":
            self.cfg.gui_width -= self.cfg.reader_lines

        # Create the message window. This could arguably be crammed into the cfg
        # class itself, however, for logging and messaging, the cfg class is
        # basically just a glorified way to do away with globals =P

        self.cfg.msg = curses.newwin(self.cfg.msg_height,\
                self.cfg.width, self.cfg.height - self.cfg.msg_height, 0)
        self.cfg.msg.bkgdset(curses.color_pair(1))
        self.cfg.msg.erase()
        self.cfg.msg.refresh()

        # Perform the main update update.
        self.gui.refresh()
        self.gui.draw_elements()

    # This also doesn't follow good signal practices, but it's an exception
    # because the backtrace is important.

    def debug_out(self, a, b):
        f = open("canto_debug_out", "w")
        for l in traceback.format_stack():
            f.write(l)
        f.close()
